src/execution_plan/ops/op_node_by_label_scan.c

```diff
@@ -13,21 +13,36 @@
 static OpResult NodeByLabelScanInit(OpBase *opBase);
 static Record NodeByLabelScanConsume(OpBase *opBase);
 static Record NodeByLabelScanConsumeFromChild(OpBase *opBase);
+static Record NodeByLabelScanNoOp(OpBase *opBase);
 static OpResult NodeByLabelScanReset(OpBase *opBase);
 static void NodeByLabelScanFree(OpBase *opBase);
 
 static inline int NodeByLabelScanToString(const OpBase *ctx, char *buf, uint buf_len) {
 	return ScanToString(ctx, buf, buf_len, ((const NodeByLabelScan *)ctx)->n);
 }
 
+/* Upon every call to create or reset iterator, there is need to see if the matrix size might have changed.
+ * The upper bound of the range should be a result of comparing the INPUT upper bound, and the maximal number of rows in matrix.*/
+static inline void _setIndices(NodeByLabelScan *op) {
+	// The largest possible entity ID is the same as Graph_RequiredMatrixDim.
+	NodeID max_id;
+	if(op->id_range) max_id = op->id_range->include_max ? op->id_range->max : op->id_range->max - 1;
+	else max_id = op->maxId;
+	op->maxId = MIN(max_id,  Graph_RequiredMatrixDim(op->g) - 1);
+}
+
 OpBase *NewNodeByLabelScanOp(const ExecutionPlan *plan, const QGNode *n) {
 	NodeByLabelScan *op = malloc(sizeof(NodeByLabelScan));
 	GraphContext *gc = QueryCtx_GetGraphCtx();
 	op->g = gc->g;
 	op->n = n;
 	op->iter = NULL;
-	op->_zero_matrix = NULL;
 	op->child_record = NULL;
+	op->id_range = NULL;
+
+	// The smallest possible entity ID is 0.
+	op->minId = 0;
+	op->maxId = UINT64_MAX;
```

Comment:
we should not modify the given range object since the max index is evaluated against matrix size. moreover, I thought it will be more informative to have it as a "read-only" object so it will not be mutated throughout execution, for debug convenience.